packages/network-controller/src/provider-api-tests/shared-tests.ts

```diff
@@ -0,0 +1,2673 @@
+/* eslint-disable jest/require-top-level-describe, jest/no-export, jest/no-identical-title, jest/no-if */
+
+import { fill } from 'lodash';
+import {
+  buildMockParams,
+  buildRequestWithReplacedBlockParam,
+  waitForNextBlockTracker,
+  waitForPromiseToBeFulfilledAfterRunningAllTimers,
+  withMockedCommunications,
+  withNetworkClient,
+} from './helpers';
+import { testsForRpcMethodAssumingNoBlockParam } from './no-block-param';
+
+export const buildInfuraClientRetriesExhaustedErrorMessage = (
+  reason: string,
+) => {
+  return new RegExp(
+    `^InfuraProvider - cannot complete request. All retries exhausted\\..+${reason}`,
+    'us',
+  );
+};
+
+export const buildFetchFailedErrorMessage = (url: string, reason: string) => {
+  return new RegExp(
+    `^request to ${url}(/[^/ ]*)+ failed, reason: ${reason}`,
+    'us',
+  );
+};
+
+type TestsForRpcMethodNotHandledByMiddlewareOptions = {
+  providerType: 'infura' | 'custom';
+  numberOfParameters: number;
+};
+
+export const testsForRpcMethodNotHandledByMiddleware = (
+  method: string,
+  {
+    providerType,
+    numberOfParameters,
+  }: TestsForRpcMethodNotHandledByMiddlewareOptions,
+) => {
+  if (providerType !== 'infura' && providerType !== 'custom') {
+    throw new Error(
+      `providerType must be either "infura" or "custom", was "${providerType}" instead`,
+    );
+  }
+
+  it('attempts to pass the request off to the RPC endpoint', async () => {
+    const request = {
+      method,
+      params: fill(Array(numberOfParameters), 'some value'),
+    };
+    const expectedResult = 'the result';
+
+    await withMockedCommunications({ providerType }, async (comms) => {
+      comms.mockNextBlockTrackerRequest({ blockNumber: '0x1' });
+      comms.mockRpcCall({
+        request,
+        response: { result: expectedResult },
+      });
+      const actualResult = await withNetworkClient(
+        { providerType },
+        ({ makeRpcCall }) => makeRpcCall(request),
+      );
+
+      expect(actualResult).toStrictEqual(expectedResult);
+    });
+  });
+};
+
+type TestsForRpcMethodWithStaticResult = {
+  providerType: 'infura' | 'custom';
+  numberOfParameters: number;
+  result: any;
+};
+
+export const testsForRpcMethodWithStaticResult = (
+  method: string,
+  {
+    providerType,
+    numberOfParameters,
+    result,
+  }: TestsForRpcMethodWithStaticResult,
+) => {
+  if (providerType !== 'infura' && providerType !== 'custom') {
+    throw new Error(
+      `providerType must be either "infura" or "custom", was "${providerType}" instead`,
+    );
+  }
+
+  it('attempts to pass the request off to the RPC endpoint', async () => {
```

Comment:
Nit: This test name might need to be changed if it never actually hits the RPC endpoint.